File: packages/js-dpp/test/unit/identity/stateTransition/IdentityUpdateTransition/IdentityUpdateTransition.spec.js
const IdentityPublicKey = require('../../../../../lib/identity/IdentityPublicKey');

const stateTransitionTypes = require(
  '../../../../../lib/stateTransition/stateTransitionTypes',
);

const protocolVersion = require('../../../../../lib/version/protocolVersion');
const Identifier = require('../../../../../lib/identifier/Identifier');

const getIdentityUpdateTransitionFixture = require('../../../../../lib/test/fixtures/getIdentityUpdateTransitionFixture');
const generateRandomIdentifier = require('../../../../../lib/test/utils/generateRandomIdentifier');

describe('IdentityUpdateTransition', () => {
  let rawStateTransition;
  let stateTransition;

  beforeEach(() => {
    stateTransition = getIdentityUpdateTransitionFixture();
    rawStateTransition = stateTransition.toObject();
  });

  describe('#getType', () => {
    it('should return IDENTITY_UPDATE type', () => {
      expect(stateTransition.getType()).to.equal(stateTransitionTypes.IDENTITY_UPDATE);
    });
  });

  describe('#setIdentityId', () => {
    it('should set identityId', () => {
      const id = generateRandomIdentifier();

      stateTransition.setIdentityId(id);

      expect(stateTransition.identityId).to.deep.equal(id);
    });
  });

  describe('#getIdentityId', () => {
    it('should return identityId', () => {
      expect(stateTransition.getIdentityId()).to.deep.equal(rawStateTransition.identityId);
    });
  });

  describe('#getRevision', () => {
    it('should return revision', () => {
      expect(stateTransition.getRevision()).to.equal(rawStateTransition.revision);
    });
  });

  describe('#setRevision', () => {
    it('should set revision', () => {
      stateTransition.setRevision(42);

      expect(stateTransition.revision).to.equal(42);
    });
  });

  describe('#getOwnerId', () => {
    it('should return owner id', () => {
      expect(stateTransition.getOwnerId()).to.deep.equal(
        rawStateTransition.identityId,
      );
    });
  });

  describe('#getPublicKeysToAdd', () => {
    it('should return public keys to add', () => {
      expect(stateTransition.getPublicKeysToAdd()).to.deep.equal(
        rawStateTransition.addPublicKeys.map((rawPublicKey) => new IdentityPublicKey(rawPublicKey)),
      );
    });
  });

  describe('#setPublicKeysToAdd', () => {
    it('should set public keys to add', () => {
      const publicKeys = [new IdentityPublicKey({
        id: 0,
        type: IdentityPublicKey.TYPES.BLS12_381,
        purpose: 0,
        securityLevel: 0,
        readOnly: true,
        data: Buffer.from('01fac99ca2c8f39c286717c213e190aba4b7af76db320ec43f479b7d9a2012313a0ae59ca576edf801444bc694686694', 'hex'),
      })];

      stateTransition.setPublicKeysToAdd(publicKeys);

      expect(stateTransition.addPublicKeys).to.have.deep.members(publicKeys);
    });
  });

  describe('#getPublicKeyIdsToDisable', () => {
    it('should return public key ids to disable', () => {
      expect(stateTransition.getPublicKeyIdsToDisable())
        .to.deep.equal(stateTransition.disablePublicKeys);
    });
  });

  describe('#setPublicKeyIdsToDisable', () => {
    it('should set public key ids to disable', () => {
      stateTransition.setPublicKeyIdsToDisable([1, 2]);

      expect(stateTransition.disablePublicKeys).to.deep.equal([1, 2]);
    });
  });

  describe('#getPublicKeysDisabledAt', () => {
    it('should return time to disable public keys', () => {
      expect(stateTransition.getPublicKeysDisabledAt())
        .to.deep.equal(new Date(stateTransition.publicKeysDisabledAt));
    });
  });

  describe('#setPublicKeysDisabledAt', () => {
    it('should set time to disable public keys', () => {
      const now = new Date();

      stateTransition.setPublicKeysDisabledAt(now);

      expect(stateTransition.publicKeysDisabledAt).to.deep.equal(new Date(now));
    });
  });

  describe('#toObject', () => {
    it('should return raw state transition', () => {
      rawStateTransition = stateTransition.toObject();

      expect(rawStateTransition).to.deep.equal({
        protocolVersion: protocolVersion.latestVersion,
        type: stateTransitionTypes.IDENTITY_UPDATE,
        signature: undefined,
        identityId: rawStateTransition.identityId,
        revision: rawStateTransition.revision,
        publicKeysDisabledAt: rawStateTransition.publicKeysDisabledAt,
        addPublicKeys: rawStateTransition.addPublicKeys,
        disablePublicKeys: rawStateTransition.disablePublicKeys,
        signaturePublicKeyId: undefined,
      });
    });

    it('should return raw state transition without signature', () => {
      rawStateTransition = stateTransition.toObject({ skipSignature: true });

      expect(rawStateTransition).to.deep.equal({
        protocolVersion: protocolVersion.latestVersion,
        type: stateTransitionTypes.IDENTITY_UPDATE,
        identityId: rawStateTransition.identityId,
        revision: rawStateTransition.revision,
        publicKeysDisabledAt: rawStateTransition.publicKeysDisabledAt,
        addPublicKeys: rawStateTransition.addPublicKeys,
        disablePublicKeys: rawStateTransition.disablePublicKeys,
      });
    });

    it('should return raw state transition without optional properties', () => {
      stateTransition.setPublicKeyIdsToDisable(undefined);
      stateTransition.setPublicKeysDisabledAt(undefined);
      stateTransition.setPublicKeysToAdd(undefined);

      rawStateTransition = stateTransition.toObject();

      expect(rawStateTransition).to.deep.equal({
        protocolVersion: protocolVersion.latestVersion,
        type: stateTransitionTypes.IDENTITY_UPDATE,
        signature: undefined,
        identityId: rawStateTransition.identityId,
        revision: rawStateTransition.revision,
        signaturePublicKeyId: undefined,
      });
    });
  });

  describe('#toJSON ', () => {
    it('should return JSON representation of state transition', () => {
      const jsonStateTransition = stateTransition.toJSON();

      expect(jsonStateTransition).to.deep.equal({
        protocolVersion: protocolVersion.latestVersion,
        type: stateTransitionTypes.IDENTITY_UPDATE,
        signature: undefined,
        identityId: stateTransition.getIdentityId().toString(),
        revision: rawStateTransition.revision,
        publicKeysDisabledAt: rawStateTransition.publicKeysDisabledAt,
        addPublicKeys: stateTransition.getPublicKeysToAdd().map((k) => k.toJSON()),
        disablePublicKeys: rawStateTransition.disablePublicKeys,
        signaturePublicKeyId: undefined,
      });
    });
  });

  describe('#getModifiedDataIds', () => {
    it('should return ids of topped up identity', () => {
      const result = stateTransition.getModifiedDataIds();

      expect(result.length).to.be.equal(1);
      const identityId = result[0];

      expect(identityId).to.be.an.instanceOf(Identifier);
      expect(identityId).to.be.deep.equal(rawStateTransition.identityId);
    });
  });

  describe('#isDataContractStateTransition', () => {
    it('should return false', () => {
      expect(stateTransition.isDataContractStateTransition()).to.be.false();
    });
  });

  describe('#isDocumentStateTransition', () => {
    it('should return false', () => {
      expect(stateTransition.isDocumentStateTransition()).to.be.false();
    });
  });

  describe('#isIdentityStateTransition', () => {
    it('should return true', () => {
      expect(stateTransition.isIdentityStateTransition()).to.be.true();
    });
  });
});

File: packages/rs-dpp/src/tests/identity/state_transition/identity_update_transition/identity_update_transition_spec.rs
use chrono::Utc;
use serde_json::{json, Value as JsonValue};

use crate::{
    identity::{
        state_transition::identity_update_transition::identity_update_transition::IdentityUpdateTransition,
        KeyType, Purpose, SecurityLevel,
    },
    prelude::IdentityPublicKey,
    state_transition::{
        StateTransitionConvert, StateTransitionIdentitySigned, StateTransitionType,
    },
    tests::{
        fixtures::get_identity_update_transition_fixture, utils::generate_random_identifier_struct,
    },
    util::string_encoding::Encoding,
};

struct TestData {
    transition: IdentityUpdateTransition,
    raw_transition: JsonValue,
}

fn setup_test() -> TestData {
    let transition = get_identity_update_transition_fixture();
    let raw_transition = transition.to_object(false).unwrap();
    TestData {
        transition,
        raw_transition,
    }
}

#[test]
fn get_type() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(StateTransitionType::IdentityUpdate, transition.get_type());
}

#[test]
fn set_identity_id() {
    let TestData { mut transition, .. } = setup_test();
    let id = generate_random_identifier_struct();
    transition.set_identity_id(id.clone());
    assert_eq!(&id, transition.get_identity_id());
}

#[test]
fn get_revision() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(0, transition.get_revision());
}

#[test]
fn set_revision() {
    let TestData { mut transition, .. } = setup_test();
    transition.set_revision(42);
    assert_eq!(42, transition.get_revision());
}

#[test]
fn get_owner_id() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(&transition.identity_id, transition.get_owner_id());
}

#[test]
fn get_public_keys_to_add() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(
        &transition.add_public_keys,
        transition.get_public_keys_to_add()
    );
}

#[test]
fn set_public_keys_to_add() {
    let TestData { mut transition, .. } = setup_test();
    let id_public_key = IdentityPublicKey {
						id : 0,
            key_type: KeyType::BLS12_381,
            purpose: Purpose::AUTHENTICATION,
            security_level : SecurityLevel::CRITICAL,
            read_only: true,
            data: hex::decode("01fac99ca2c8f39c286717c213e190aba4b7af76db320ec43f479b7d9a2012313a0ae59ca576edf801444bc694686694").unwrap(),
            disabled_at : None,
        };
    transition.set_public_keys_to_add(vec![id_public_key.clone()]);

    assert_eq!(vec![id_public_key], transition.get_public_keys_to_add());
}

#[test]
fn get_disable_public_keys() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(
        transition.disable_public_keys,
        transition.get_public_key_ids_to_disable()
    );
}

#[test]
fn set_disable_public_keys() {
    let TestData { mut transition, .. } = setup_test();
    let id_to_disable = vec![1, 2];
    transition.set_public_key_ids_to_disable(id_to_disable.clone());

    assert_eq!(&id_to_disable, transition.get_public_key_ids_to_disable());
}

#[test]
fn get_public_key_disabled_at() {
    let TestData { transition, .. } = setup_test();
    assert_eq!(
        transition.public_keys_disabled_at,
        transition.get_public_keys_disabled_at()
    );
}

#[test]
fn set_public_key_disabled_at() {
    let TestData { mut transition, .. } = setup_test();
    let now = Utc::now().timestamp_millis() as u64;
    transition.set_public_keys_disabled_at(Some(now));

    assert_eq!(Some(now), transition.get_public_keys_disabled_at());
}

#[test]
fn to_object() {
    let TestData { transition, .. } = setup_test();
    let result = transition
        .to_object(false)
        .expect("conversion to object shouldn't fail");

    let expected_raw_state_transition = json!({
        "protocolVersion" : 1,
        "type" : 5,
        "signature" : [],
        "signaturePublicKeyId": 0,
        "identityId" : transition.identity_id.to_buffer(),
        "revision": 0,
        "disablePublicKeys" : [0],
        "publicKeysDisabledAt" : 1234567,
        "addPublicKeys" : [
            {

                "id" : 3,
                "purpose" : 0,
                "type": 0,
                "securityLevel" : 0,
                "data" :base64::decode("AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH").unwrap(),
                "readOnly" : false
            }
        ]
    });

    assert_eq!(expected_raw_state_transition, result);
}

#[test]
fn to_object_with_signature_skipped() {
    let TestData { transition, .. } = setup_test();
    let result = transition
        .to_object(true)
        .expect("conversion to object shouldn't fail");

    let expected_raw_state_transition = json!({
        "protocolVersion" : 1,
        "type" : 5,
        "signaturePublicKeyId": 0,
        "identityId" : transition.identity_id.to_buffer(),
        "revision": 0,
        "disablePublicKeys" : [0],
        "publicKeysDisabledAt" : 1234567,
        "addPublicKeys" : [
            {

                "id" : 3,
                "purpose" : 0,
                "type": 0,
                "securityLevel" : 0,
                "data" :base64::decode("AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH").unwrap(),
                "readOnly" : false
            }
        ]
    });

    assert_eq!(expected_raw_state_transition, result);
}

#[test]
fn to_json() {
    let TestData { transition, .. } = setup_test();
    let result = transition
        .to_json()
        .expect("conversion to json shouldn't fail");

    let expected_raw_state_transition = json!({
        "protocolVersion" : 1,
        "type" : 5,
        "signature" : "",
        "signaturePublicKeyId": 0,
        "identityId" : transition.identity_id.to_string(Encoding::Base58),
        "revision": 0,
        "disablePublicKeys" : [0],
        "publicKeysDisabledAt" : 1234567,
        "addPublicKeys" : [
            {

                "id" : 3,
                "purpose" : 0,
                "type": 0,
                "securityLevel" : 0,
                "data" : "AkVuTKyF3YgKLAQlLEtaUL2HTditwGILfWUVqjzYnIgH",
                "readOnly" : false
            }
        ]
    });

    assert_eq!(expected_raw_state_transition, result);
}
